### ISSUE TITLE
Support getting max microversion in serviceclient

### ISSUE THREAD

**Issue by openstacker** [gophercloud/gophercloud#1520]
Background:

I'm triying to consume a new API support introduced in gophercloud recently. Now in the consumer of the gopthercloud, I'd like to check the serviceclient version firstly before using the new API. If it doesn't support, then I can use other alternative way.

For example, Magnum recently introduced the new /resize API to resize a COE cluster. Now in Kubernetes Cluster Autoscaler, I'd like to check if the service client can support the new API before calling. Otherwise, I can switch to another way so that the driver can work for different Magnum environment.

Proposal:

A new field "MaxMicroVersion" (we can also have the MinMicroVersion as well) can be added here https://github.com/gophercloud/gophercloud/blob/master/service_client.go#L30

New function like https://github.com/gophercloud/gophercloud/blob/master/service_client.go#L38 can be created to expose the version.

**Comment by openstacker** [comment]
@jtopjian @ozerovandrei Could you please comment this? Thanks.

**Comment by jtopjian** [comment]
There's been discussion about this in the past. In short, it's not an easy thing to do and would require a good amount of work.

The absolute minimum required to auto-detect the microversions for Magnum (and not have any concern for the rest of Gophercoud), is for someone to write an `apiversions` package for Gophercloud's Magnum support, similar to [`Manila's`](https://github.com/gophercloud/gophercloud/tree/master/openstack/sharedfilesystems/apiversions).

Once that's done, your Gophercloud-enabled application should be able to use that package to query for the supported versions and make an informed decision there.

An easier option is to make microversion-specific features opt-in by the user. For example: https://github.com/terraform-providers/terraform-provider-openstack/blob/2b58e07318321a06def676f499e8056dc2811f81/openstack/resource_openstack_compute_volume_attach_v2.go#L86-L89

So in your case, if there's the possibility to create a flag of `use_magnum_autoscale = true` or something like that, that's the best way to go.

**Comment by openstacker** [comment]
Thanks @jtopjian  I was thinking to have a high level, general solution for this. But if you think an api version package for Mganum is better, then I'm happy to go for that.

As the "easier option" you mentioned in terraform, that's could be easier for Terraform, because Terraform just pass the microversion to OpenStack. But for our case, personally, I prefer to hide the details and just avoid another config option.

Thank you very much for your quick response.

**Comment by jtopjian** [comment]
> Thanks @jtopjian I was thinking to have a high level, general solution for this. But if you think an api version package for Mganum is better, then I'm happy to go for that.

Even if you wanted to integrate this into Gophercloud core, each service will need an `apiversions` package in order to query the endpoints. This is the main reason this feature hasn't been integrated into Gophercloud. Each service presents the API versions slightly differently and no one has gotten around to writing a package for each service.

**Comment by openstacker** [comment]
I see the point. I'm going to follow the same way like Manila to add the new apiversions package, and I would encourage you advice all the new comers like me to follow the consistent way so that in the future we can get the Max/Min Microversion support at the common level. Cheers.

**Comment by jtopjian** [comment]
> I would encourage you advice all the new comers like me to follow the consistent way so that in the future we can get the Max/Min Microversion support at the common level

Well, it'd be nice if all the OpenStack services did _anything_ consistent to help SDKs follow consistent paths 😉

**Comment by openstacker** [comment]
Hah, that's one of things I mentioned in my TC self nomination email. OpenStack did a bad job for such kind of minor things. From a user perspective, all the Service APIs should be consistent.

**Comment by jtopjian** [comment]
I agree. Don't get me wrong - I've read all of the specs for Microversions, Service Discovery, etc and I agree with those specs. The difficulty is not just getting all services to conform to those specs, but for SDKs to also support _prior_ behavior so the majority of deployed production clouds are still supported. To have Gophercloud support all of that takes a lot of time.

But anyway, creating an `apiversions` package for Magnum will get you moving.

One other thing to keep in mind: microversions must be of type "string". You will need to split the major and minor and convert each to an integer.

This is because a version like "3.10" means major version 3 and minor version 10. Converting "3.10" to a float will change it to "3.1".

**Comment by openstacker** [comment]
Gotcha, thanks for all the good comments. Working with gophercloud team is always like a breeze. Cheers.
